### DISCUSSION TITLE
vivado synthesis failed of a CNN model

### DISCUSSION THREAD

**Discussion by brtsfr** [fastmachinelearning/hls4ml#1025] [Q&A] [answered]
hi all, i am a new learner here. Today i was tring to load a torch CNN model, and transfer it to FPGA firmware. i managed to finish other steps, until i got this error. please see log.txt for the full information showed in terminal. Also i attached .py file that i'm running.
[log.txt](https://github.com/user-attachments/files/16027018/log.txt)
[test.py.txt](https://github.com/user-attachments/files/16027033/test.py.txt)
please help :) many thanks.

**Comment by brtsfr** [comment]
when i look into the log.txt, i found the first error is that: ../../../../firmware/my_hls_projectCNN.cpp:49:5: error: 'conv_2d_cf' is not a member of 'nnet'.

**Reply by brtsfr** [reply]
and i noticed that in Issue#502, Conv1d compile error. "‘conv_1d_cf’ is not a member of ‘nnet’; did you mean ‘conv_1d_cl’?" #502, it says channels_first will not be supported. So does this mean that i need to modify codes in convolution.py and pooling.py of hls4ml lib to channels_last, or can i add a conv_2d_cf definition somewhere?

**Reply by brtsfr** [reply]
because i found some nnet_conv2d.h files which contain conv_2d_cf.

**Reply by brtsfr** [reply]
update: i found some .h files at https://github.com/YanLunHuang/conv2d_test.git, and managed to solve those errors, until i meet SIGSEGV.

**Comment by vloncar** [comment, marked as answer]
hls4ml implements CNN layers expecting a "channels last" format (also known as `(N, H, W, C)`) which is used by Keras/TF. PyTorch by default uses "channels first" format (or `(N, C, H, W)` in other terminology. These use the suffix `_cf` in the code and are not implemented. You can instruct hls4ml to convert to "channels last" and expect input data to be in that format (and optionally the output if needed). Check the options `InputsChannelLast` and `TransposeOutputs` of the `config_from_pytorch_model`. Also, you're likely using older version of hls4ml, try installing the main branch (soon to become 1.0). Mixing and matching random .h files from various forks will likely not work :smile:

**Reply by brtsfr** [reply]
ok, i installed 0.9.0.dev38+g02688f1d. And i will try 'config_from_pytorch_model' this submodule. Thanks for help, i'm a hardware engineer of PCB design and FPGA development, so usually do it in wrong way, i appreciate your kindly explaination.

**Reply by brtsfr** [reply]
Thanks, it's working! 
 i have a new situation, HLS stuck at " Starting code transformations..." for a whole night that i have to terminate it. Here attached the screen output log and my python code. How can i solve this? Do you have any suggestions
[log.txt](https://github.com/user-attachments/files/16060685/log.txt)
[test.py.txt](https://github.com/user-attachments/files/16060686/test.py.txt)

**Reply by vloncar** [reply]
Your model is too large. 512 filters won't work. The activation tensors get too large (512*8*8) to be partitioned.

**Reply by brtsfr** [reply]
got it. let me modify the model structure.

**Reply by brtsfr** [reply]
i simplified my model structure, which is showed in model.txt. During the process i got an error: Unrolling loop completely with a factor of 288, ERROR: [XFORM 203-504] Stop unrolling loop,  because it may cause large runtime and excessive memory usage due to increase in code size. Please avoid unrolling the loop or form sub-functions for code in the loop body. ERROR: [HLS 200-70] Pre-synthesis failed. Detailed log is attached in log.txt. Also my py file is attached.  i guess it's about default_precision and default_reuse_factor, after i changed these two parameters, the error remains still. Thanks for teaching me.
[log.txt](https://github.com/user-attachments/files/16068202/log.txt)
[model.txt](https://github.com/user-attachments/files/16068203/model.txt)
[test.py.txt](https://github.com/user-attachments/files/16068204/test.py.txt)

**Reply by brtsfr** [reply]
ok, i found info here, 
https://github.com/fastmachinelearning/hls4ml/issues/1013
